apps/backend/src/auth/auth.controller.ts

```diff
@@ -5,12 +5,50 @@ import githubAuth from "./github.auth.controller";
 import localAuth from "./local.auth.controller";
 import googleAuth from "./google.auth.controller";
 import { schemaDetail } from "./auth.model";
+import { UnauthorizedError } from "../errors";
+import { prisma } from "@nutrishare/db";
+import authService from "./auth.service";
 
 export default new Elysia({ prefix: "/auth" })
   .use(userModel)
+  .use(authService)
   .use(localAuth)
   .use(githubAuth)
   .use(googleAuth)
+  .post(
+    "/refresh",
+    async ({ body: { refreshToken: providedRefreshToken }, authService }) => {
+      const refreshToken = await authService.verifyRefreshToken(
+        providedRefreshToken,
+      );
+      const userId = refreshToken.id;
+
+      const expiredToken = await prisma.refreshToken.findFirst({
+        where: { refreshToken: providedRefreshToken, expired: true },
+      });
+      // If the used refresh token is expired, invalidate all valid refresh tokens for this user
```

Comment:
TBH, I don't really understand what's this for. Users can't have two independent sessions? Do you have any resources about this practice?

[2]

Reply:
I learned about this practice from [Auth0's article on refresh tokens](https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/#Refresh-Token-Automatic-Reuse-Detection). If an attacker obtains your refresh token and tries to use it later in a reply attack, we invalidate all existing refresh tokens just in case your device is compromised. The only thing this prevents is refresh attacks, but I think it's a good security measure with not too much additional cost.

> Users can't have two independent sessions?
This is actually a very good point. My implementation is very naive, as it invalidates all tokens of the user, not just a token "family".

One way to solve this and implement actual refresh token families would be to add a `family` column to the `RefreshToken` model. It would be randomized when the user first authenticates (e.g. /login) and saved in the JWT, so we can later identify which family the token comes from. If a refresh token is reused, invalidate all tokens of the same family, not the same user. This would allow one user to have multiple "sessions" without disruptions.

Reply:
Cool, that makes sense :)

Reply:
Yeah, maybe this could be improved later. Not that big of an issue IMO. This is more of an UX inconvenience than a security issue.